src/backend/distributed/utils/datautils.c

```diff
@@ -0,0 +1,32 @@
+/*-------------------------------------------------------------------------
+ *
+ * datautils.c
+ *
+ * This file contains functions to perform useful operations on data types.
+ *
+ * Copyright (c) 2014-2016, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+#include "c.h"
+
+#include "distributed/datautils.h"
+
+
+/*
+ * Allocates a new uint64 on the heap and copies the input to that space,
+ * returning a pointer to allocated memory. Useful for lists with 64-bit
+ * integers.
+ */
+uint64 *
+AllocateUint64(uint64 value)
+{
+	uint64 *allocatedValue = (uint64 *) palloc0(sizeof(uint64));
```

Comment:
On 2016-07-13 14:32:57 -0700, Jason Petersen wrote:
> > +
> > +#include "postgres.h"
> > +#include "c.h"
> > +
> > +#include "distributed/datautils.h"
> > +
> > +
> > +/*
> > + * Allocates a new uint64 on the heap and copies the input to that space,
> > + * returning a pointer to allocated memory. Useful for lists with 64-bit
> > + * integers.
> > + */
> > +uint64 *
> > +AllocateUint64(uint64 value)
> > +{
> > +	uint64 *allocatedValue = (uint64 *) palloc0(sizeof(uint64));
> 
> (a) PostgreSQL doesn't have unsigned methods like that, no?

It does:

/*
 * DatumGetUInt64
 *		Returns 64-bit unsigned integer value of a datum.
 *
 * Note: this macro hides whether int64 is pass by value or by reference.
 */

#ifdef USE_FLOAT8_BYVAL
#define DatumGetUInt64(X) ((uint64) GET_8_BYTES(X))
#else
#define DatumGetUInt64(X) (* ((uint64 *) DatumGetPointer(X)))
#endif
/*
 * UInt64GetDatum
 *		Returns datum representation for a 64-bit unsigned integer.
 *
 * Note: if int64 is pass by reference, this function returns a reference
 * to palloc'd space.
 */

#ifdef USE_FLOAT8_BYVAL
#define UInt64GetDatum(X) ((Datum) SET_8_BYTES(X))
#else
#define UInt64GetDatum(X) Int64GetDatum((int64) (X))
#endif

> (b) I don't quite understand how this would help here… I _need_ to allocate a `uint64` in the `TopTransactionContext`, so it can be added to a list and live there in that context. If it were a smaller `int`, I could just put it in an `int` list, but there's nowhere to put a 64-bit value in a list, even it can be pass-by-value on a certain platform.

You can just use a non-int list. Datums are guaranteed to be storable in
a size_t wide variable.

Andres